### PR TITLE
fix(fx): make all `make_fx` invocations isolated (opaque to higher `make_fx` invocations) by default

### LINKED ISSUE #88996: Nested FX tracing doesn't work when outer tracing mode is symbolic
### 🐛 Describe the bug

nvFuser executor for AOT Autograd + PrimTorch uses isolated make_fx to get decomposed subgraphs for analysis. This doesn't work if the outer tracing mode is symbolic, it fails with the following error:
```py
File ~/dev/pytorch/master/torch/fx/experimental/proxy_tensor.py:87, in get_proxy_slot(obj, tracer, default, transform)
     85 if tracer not in d:
     86     if default is no_default:
---> 87         raise KeyError(f"{obj} is not tracked with proxy for {tracer}")
     88     else:
     89         return default

KeyError: 's0 is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7facce5006d0>'
```
Here's code to reproduce the failure:
```python
import torch
from torch.fx.experimental.proxy_tensor import make_fx, wrapper_and_args_for_make_fx

def func(a, b):
    return b.expand([1, a.shape[0], b.shape[-1]])

a = torch.randn(3, 4, device="cuda")
b = torch.randn(4, device="cuda")

class TestMode(torch.overrides.TorchFunctionMode):
    def __torch_function__(self, func, types, args=(), kwargs=None):
        if kwargs is None:
            kwargs = {}

        if torch.overrides.resolve_name(func) in ["torch.Tensor.expand"]:
            print(f"TestMode: {func} {args} {kwargs}")
            wrapped, all_args = wrapper_and_args_for_make_fx(func, args, kwargs)
            gm = make_fx(wrapped, tracing_mode="real")(all_args)
            gm.graph.print_tabular()

        return func(*args, **kwargs)

with TestMode():
    gm = make_fx(func, tracing_mode="symbolic")(a, b)

gm.graph.print_tabular()
```

### Versions

Latest master.

cc @ezyang @SherlockNoMad @soumith @EikanWang @jgong5 @wenzhe-nrv

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/issues/88996#issuecomment-1409174554

Example code:
```python
import torch
from torch.fx.experimental.proxy_tensor import make_fx, wrapper_and_args_for_make_fx

@torch.fx.wrap
def func(a, b):
    return b.expand([1, a.shape[0], b.shape[-1]])

a = torch.randn(3, 4)
b = torch.randn(4)

class TestMode(torch.overrides.TorchFunctionMode):
    def __torch_function__(self, func, types, args=(), kwargs={}):
        if torch.overrides.resolve_name(func) in ["torch.Tensor.expand"]:
            print(f"TestMode: {func} {args} {kwargs}")
            wrapped, all_args = wrapper_and_args_for_make_fx(func, args, kwargs)
            gm = make_fx(wrapped, tracing_mode="real")(all_args)

        return func(*args, **kwargs)

with TestMode():
    gm = make_fx(func, tracing_mode="symbolic")(a, b)

gm.graph.print_tabular()
```
Before:
```
opcode         name        target               args                              kwargs
-------------  ----------  -------------------  --------------------------------  --------
placeholder    a_1         a_1                  ()                                {}
placeholder    b_1         b_1                  ()                                {}
call_function  detach      aten.detach.default  (b_1,)                            {}
call_function  detach_1    aten.detach.default  (detach,)                         {}
call_function  sym_size    aten.sym_size        (a_1, 0)                          {}
call_function  sym_size_1  aten.sym_size        (b_1, 0)                          {}
call_function  expand      aten.expand.default  (b_1, [1, sym_size, sym_size_1])  {}
call_function  detach_2    aten.detach.default  (expand,)                         {}
call_function  expand_1    aten.expand.default  (b_1, [1, sym_size, sym_size_1])  {}
output         output      output               (expand_1,)                       {}
```

After:
```
opcode         name        target               args                              kwargs
-------------  ----------  -------------------  --------------------------------  --------
placeholder    a_1         a_1                  ()                                {}
placeholder    b_1         b_1                  ()                                {}
call_function  sym_size    aten.sym_size        (a_1, 0)                          {}
call_function  sym_size_1  aten.sym_size        (b_1, 0)                          {}
call_function  expand      aten.expand.default  (b_1, [1, sym_size, sym_size_1])  {}
output         output      output               (expand_1,)                       {}
```

### CODE REVIEW

**Comment by ezyang** [comment]
I... guess we can do this? It kind of feels better to not by default (and let get_isolated_subgraph be used for this case) because if you do it this way, there is no way to have a single trace get recorded by multiple proxy tensor modes at once. This may seem like a weird thing to want but in fact functorch nested grad does something like this (where nested grad levels get recorded onto multiple tapes one per grad level). 

The countervailing argument is make_fx doesn't return its outputs, so therefore it is not true compute and shouldn't get traced. I could be convinced by this.

**Comment by jon-chuang** [comment]
Actually, I made a mistake. 

With the changes https://github.com/pytorch/pytorch/pull/93290/commits/e61cb65bef378bb108c21782f5f5f037f4936313, when we call `make_fx` nested in another `make_fx`, the inner make_fx` will behave exactly like `get_isolated_submodule`. So we are isolating submodules by default.

Banning all fake proxy tracers except the outermost is obviously wrong.

**Comment by jon-chuang** [comment]
> (Unless: an inner tape is used in the definition of the exterior function? But anw, we seem to be moving away from fx to aot.autograd...)

Actually, aot.autograd in compiled mode uses fake tensor proxy via `make_fx` under the hood, so technically, calling autograd in a nested way may result in some issues. 

But I don't think we would because we only ever call `make_fx` more than once as it makes no sense to create a tape of the compilation process. In fact, for aot_autograd, `make_fx` is only called in `aot_dispatch_base` and `aot_dispatch_autograd`, and they are orthogonal.

**Comment by jon-chuang** [comment]
> functorch nested grad does something like this (where nested grad levels get recorded onto multiple tapes one per grad level).

By this do you mean nesting `torch.func.grad`?

Could you point to an example? I tried constructing several examples where things might fail, including nesting `torch.func.grad` and `torch.autograd.grad` but:
1. for the former, `make_fx` is always just called once, the invocation is treated as a side-effect
2. for the latter, `autograd.grad` is actually unnestable and produces a `None` result...

There is basically no compilation pathway that has nested tapes. Because it seems that `torch.autograd.autograd` itself does not rely on tapes from `make_fx`, it uses a C++ `torch._C._ImperativeEngine`. This in itself directly produces a python code artifact.

The compilation pathway that does utilize `make_fx` does not produce an artifact with a direct functional dependence on the input function.

File: test/test_proxy_tensor.py [hunk 1]
```diff
@@ -238,6 +238,19 @@ def f2(x):
 
         traced = make_fx(f2)(torch.randn(3))
         self.assertFalse(is_any_sum(traced))
+        self.assertFalse(is_any_sigmoid(traced))
+        self.assertTrue(is_any_digamma(traced))
+
+        # Verify that the forward function of a graph module produced as a 
+        # side effect of an inner `make_fx` is still traced
+        def f3(x):
```
Comment (jon-chuang) [inline test/test_proxy_tensor.py:246]: Added a test to show that graph module artifact derived from same inputs as external trace has `forward` function that is still traced when called.

File: torch/fx/experimental/proxy_tensor.py [hunk 2]
```diff
@@ -734,6 +738,23 @@ def disable_proxy_modes_tracing():
         for proxy_mode, old in zip(proxy_tensor_modes, olds):
             proxy_mode.enable_tracing = old
 
+# Disables all proxy modes except the current, most recently instantiated one. This makes the containing
+# code opaque to all higher invocations of `make_fx`
+@contextlib.contextmanager
+def enable_current_proxy_mode_exclusive():
+    # TODO: This probably doesn't correctly also disable ProxySymDispatchMode
```
Comment (ezyang) [inline torch/fx/experimental/proxy_tensor.py:745]: You're toggling `sym_mode` so yes you're handling it correctly now

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**